signer/src/stacks/api.rs

```diff
@@ -1309,15 +1309,84 @@ where
+        // is a pre-Nakamoto block we should stop rather than error out.
+        match stacks.get_tenure(header.parent_block_id).await {
+            Ok(parent_tenure_blocks) => {
+                let parent_tenure: TenureBlockHeaders = parent_tenure_blocks.into();
+                // Log if there's a large gap in bitcoin anchor heights, 
+                // expected during testnet
+                let anchor_gap = (*tenure.anchor_block_height).saturating_sub(*parent_tenure.anchor_block_height);
+                if anchor_gap > 10 {
+                    tracing::info!(
+                        current_anchor_height = %tenure.anchor_block_height,
+                        parent_anchor_height = %parent_tenure.anchor_block_height,
+                        gap = %anchor_gap,
+                        "detected large gap in bitcoin anchor heights during backfill"
+                    );
+                    // TODO: this is where we'll manually massage / insert blocks if in testnet
+                    // Increment a metric for monitoring
+                    metrics::counter!(
+                        "stacks_backfill_anchor_gaps_total",
+                        "gap_size" => if anchor_gap > 100 { "large" } else { "medium" }
+                    ).increment(1);
+                }
+                headers.push(parent_tenure);
+            }
+            Err(error) => {
+                // Check if it's a pre-Nakamoto block error
+                if is_pre_nakamoto_block_error(&error) {
+                    tracing::warn!(
+                        parent_block_id = %header.parent_block_id,
+                        current_anchor_height = %tenure.anchor_block_height,
+                        error = %error,
+                        "parent block appears to be pre-Nakamoto, stop backfill"
+                    );
+
+                    // Increment metric for monitoring pre-Nakamoto stops
+                    metrics::counter!(
+                        "stacks_backfill_pre_nakamoto_stops_total"
+                    ).increment(1);
+
+                    break;
+                } else {
+                    // Propogate other errors
+                    tracing::error!(
+                        parent_block_id = %header.parent_block_id,
+                        error = %error,
+                        "failed to fetch parent tenure during backfill"
+                    );
+                    return Err(error);
+                }
+            }
+        }
     }
 
     headers.reverse();
     Ok(headers)
 }
 
+// Helper function for backfill testnet logic, checks if an 
+// error indicates we're trying to fetch a pre-Nakamoto block
+fn is_pre_nakamoto_block_error(error: &Error) -> bool {
+    match error {
+        // If we get a 404, it's likely because the block ID refers to a pre-Nakamoto block
+        Error::StacksNodeResponse(req_error) => {
+            req_error.status() == Some(reqwest::StatusCode::NOT_FOUND)
+        }
```

Comment:
Hmm, there must be some other check that we can do to make this whole thing tighter; right now every missing block ID is treated as if we are in pre-Nakamoto territory, which doesn't seem right.

Reply:
One think that we could do is add another variable for bootstrapping, or use the fact that this is a testnet check, since this isn't a problem on mainnet.